source/common/router/filter_config_discovery_impl.cc

```diff
@@ -0,0 +1,197 @@
+        subscription_->start();
+        init_target_.ready();
+      }) {
+  subscription_->filter_config_providers_.insert(this);
+  tls_->set([](Event::Dispatcher&) -> ThreadLocal::ThreadLocalObjectSharedPtr {
+    return std::make_shared<ThreadLocalConfig>();
+  });
+}
+
+DynamicFilterConfigProviderImpl::~DynamicFilterConfigProviderImpl() {
+  subscription_->filter_config_providers_.erase(this);
+}
+
+const std::string& DynamicFilterConfigProviderImpl::name() { return subscription_->name(); }
+
+absl::optional<Http::FilterFactoryCb> DynamicFilterConfigProviderImpl::config() {
+  return tls_->getTyped<ThreadLocalConfig>().config_;
+}
+
+void DynamicFilterConfigProviderImpl::validateConfig(
+    Server::Configuration::NamedHttpFilterConfigFactory& factory) {
+  if (factory.isTerminalFilter() && !require_terminal_) {
+    throw EnvoyException(
+        fmt::format("Error: filter config {} must not be the last in the filter chain", name()));
+  } else if (!factory.isTerminalFilter() && require_terminal_) {
+    throw EnvoyException(
+        fmt::format("Error: filter config {} must be the last in the filter chain", name()));
+  }
+}
+
+void DynamicFilterConfigProviderImpl::onConfigUpdate(Http::FilterFactoryCb config,
+                                                     const std::string&) {
+  tls_->runOnAllThreads([config](ThreadLocal::ThreadLocalObjectSharedPtr previous)
+                            -> ThreadLocal::ThreadLocalObjectSharedPtr {
+    auto prev_config = std::dynamic_pointer_cast<ThreadLocalConfig>(previous);
+    prev_config->config_ = config;
+    return previous;
+  });
+}
+
+FilterConfigSubscription::FilterConfigSubscription(
+    const envoy::config::core::v3::ConfigSource& config_source,
+    const std::string& filter_config_name, Server::Configuration::FactoryContext& factory_context,
+    const std::string& stat_prefix, FilterConfigProviderManagerImpl& filter_config_provider_manager,
+    const std::string& subscription_id)
+    : Envoy::Config::SubscriptionBase<envoy::config::core::v3::TypedExtensionConfig>(
+          envoy::config::core::v3::ApiVersion::V3,
+          factory_context.messageValidationContext().dynamicValidationVisitor(), "name"),
+      filter_config_name_(filter_config_name), factory_context_(factory_context),
+      validator_(factory_context.messageValidationContext().dynamicValidationVisitor()),
+      parent_init_target_(fmt::format("FilterConfigSubscription init {}", filter_config_name_),
+                          [this]() { local_init_manager_.initialize(local_init_watcher_); }),
+      local_init_watcher_(
+          fmt::format("FilterConfigSubscription local-init-watcher {}", filter_config_name_),
+          [this]() { parent_init_target_.ready(); }),
+      local_init_target_(
+          fmt::format("FilterConfigSubscription local-init-target {}", filter_config_name_),
+          [this]() { start(); }),
+      local_init_manager_(
```

Comment:
Why do we need a local init manager? Is the filter config provider transitively warming on something other than its own subscriptions? 

In any case, I think warming and init is pretty hard to grok from looking at code, can you add some plain text explanation of what's going on so that later readers will understand the warming relationships?

Reply:
You are right, we don't need one (I thought it was needed when I branched off RDS but it's not). Refactored to make it just an init target.